app/models/account/credit_card.rb

```diff
@@ -97,17 +97,27 @@ def delete_cc_details
 
   def unstore_credit_card!
     begin
+      # TODO: An exception here would fail the operation but should we notify provider/buyer on unsuccessful response?
       response = provider_payment_gateway.try!(:threescale_unstore, credit_card_auth_code)
-      log_gateway_response(response, "unstore [auth: #{credit_card_auth_code}]")
     rescue Account::Gateway::InvalidSettingsError => error
       Rails.logger.warn("Couldn't unstore credit card details from the payment gateway: #{error.message}")
     end
 
-    return if payment_detail.destroyed?
+    unless payment_detail.destroyed?
+      clear_cc_attributes
+    end
 
-    self.credit_card_auth_code = nil
-    self.credit_card_expires_on = nil
-    self.credit_card_partial_number = nil
+    response
+  end
+
+  private
+
+  def unstore_credit_card_on_destroy!
+    response = unstore_credit_card!
+    notify_gateway_response(response)
+  rescue => error
+    event = Accounts::CreditCardUnstoreFailedEvent.create(self, error.message)
+    Rails.application.config.event_store.publish_event(event)
```

Comment:
Instead of duplicating, it would be better to modify `notify_gateway_response` to be able to be called for both response failure and exception.

Reply:
Handling multiple data types within a method is an anti-pattern. It is cleaner code when each method handles a single data type. Of course it cannot be taken to the extreme, like DRY cannot be taken to the extreme. I applied some additional DRYing though. See last 2 commits.